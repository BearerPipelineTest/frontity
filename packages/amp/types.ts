import { Package, Action, MergePackages } from "frontity/types";
import Html2React from "@frontity/html2react/types";
import Source from "@frontity/source/types";

/**
 * The types of the AMP Frontity package.
 */
interface AMP extends Package {
  /**
   * The name of this package.
   */
  name: "@frontity/amp";

  /**
   * Actions exposed by this package.
   */
  actions: {
    /**
     * Amp namespace.
     */
    amp: {
      /**
       * An internal action that bootstraps the initialization.
       *
       * @remarks
       * This action is not meant to be run by the user, but by the Frontity
       * framework.
       */
      init: Action<Packages>;
    };
  };
}

/**
 * Packages used internally by the AMP package.
 */
export type Packages = MergePackages<AMP, Source, Html2React>;

export default AMP;
