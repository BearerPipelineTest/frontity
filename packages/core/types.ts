import { TransformOptions } from "@babel/core";
import { Configuration } from "webpack";

export type Mode = "development" | "production";
export type Target = "module" | "es5" | "server";

export interface BabelConfigs {
  module: TransformOptions;
  es5: TransformOptions;
  server: TransformOptions;
}

export interface WebpackConfigs {
  module: Configuration;
  es5: Configuration;
  server: Configuration;
}

export interface FrontityConfig {
  outDir: string;
}

export interface Config {
  babel: BabelConfigs;
  webpack: WebpackConfigs;
  frontity: FrontityConfig;
}

export interface EntryPoints {
  name: string;
  path: string;
}

export type FrontityTags = {
  script?: string;
  link?: string;
  style?: string;
};

export type HeadTags = {
  bodyAttributes: string;
  htmlAttributes: string;
  head: string[];
};

export type Template = ({
  html,
  scripts,
  head,
  htmlAttributes,
  bodyAttributes,
}: {
  html: string;
  scripts: [];
  head: [];
  htmlAttributes?: string;
  bodyAttributes?: string;
}) => string;
