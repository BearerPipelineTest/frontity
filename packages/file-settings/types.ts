/**
 * Abstract base type used to make the {@link Multi} and {@link Normalized} types.
 */
type Base = {
  /**
   * Field to distinguish between different sites.
   */
  match?: string[];

  /**
   * The mode of this site.
   *
   * Used to choose to render regular HTML (default) or AMP-compatible HTML.
   */
  mode?: "default" | "amp";

  /**
   * Global state.
   */
  state?: Record<string, unknown>;
};

/**
 * The base type of a Frontity package.
 */
export type Package = {
  /**
   * Package name.
   */
  name: string;

  /**
   * Whether the package should be active or not.
   */
  active?: boolean;

  /**
   * Package state.
   */
  state?: Record<string, unknown>;
};

/**
 * Package that has been normalized.
 */
type NormalizedPackage = Required<Package>;

/**
 * Multi-site package.
 */
type Multi = Base & {
  /**
   * Site name.
   */
  name: string;
};

/**
 * Represents normalized site settings.
 */
type Normalized<T> = Base & {
  /**
   * Mode of the site.
   */
  mode: "default" | "amp";

  /**
   * Site state.
   */
  state: Record<string, unknown>; // Default: {}

  /**
   * Packages of the site.
   */
  packages: T[];
};

/**
 * Normalized site settings.
 */
export type NormalizedSettings<T = NormalizedPackage> = Normalized<T> & Multi;

/**
 * A Frontity site.
 */
export type Site = {
  /**
   * Site name.
   */
  name: string;

  /**
   * The mode of this site.
   */
  mode: "default" | "amp";

  /**
   * Site packages.
   */
  packages: string[];
};
