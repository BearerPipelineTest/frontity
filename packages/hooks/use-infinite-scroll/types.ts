import useInView from "../use-in-view";
import { MergePackages } from "frontity/types";
import Source from "@frontity/source/types";
import Router from "@frontity/router/types";

/**
 * The intersection options passed down to `useInView`.
 *
 * They come originally from the
 * [`react-intersection-observer`](https://github.com/thebuilder/react-intersection-observer)
 * package.
 */
export type IntersectionOptions = Parameters<typeof useInView>[0];

/**
 * Options of the {@link useInfiniteScroll} hook.
 */
export type UseInfiniteScrollOptions = {
  /**
   * The current link that should be used to start the infinite scroll.
   */
  currentLink: string;

  /**
   * The next link that should be fetched and loaded once the user scrolls
   * down.
   */
  nextLink?: string;

  /**
   * The options for the internal {@link useInView} hook used for the
   * `actions.source.fetch`.
   */
  fetchInViewOptions?: IntersectionOptions;

  /**
   * The options for the internal {@link useInView} hook used for the
   * `actions.router.set`.
   */
  routeInViewOptions?: IntersectionOptions;
};

/**
 * Options of the {@link useInfiniteScroll} hook.
 */
export type UseInfiniteScrollOutput =
  | {
      /**
       * Boolean indicating if the Intersection Observer is supported or not by
       * the browser.
       */
      supported: false;

      /**
       * The ref that should be attached to the element used to trigger
       * `actions.router.set`.
       */
      routeRef?: undefined;

      /**
       * The ref that should be attached to the element used to trigger
       * `actions.source.fetch`.
       */
      fetchRef?: undefined;

      /**
       * Boolean that indicates when the element used to trigger
       * `actions.router.set` is in the screen.
       */
      routeInView?: undefined;

      /**
       * Boolean that indicates when the element used to trigger
       * `actions.source.fetch` is in the screen.
       */
      fetchInView?: undefined;
    }
  | {
      /**
       * Boolean indicating if the Intersection Observer is supported or not by
       * the browser.
       */
      supported: true;

      /**
       * The ref that should be attached to the element used to trigger
       * `actions.router.set`.
       */
      routeRef: (node?: Element) => void;

      /**
       * The ref that should be attached to the element used to trigger
       * `actions.source.fetch`.
       */
      fetchRef: (node?: Element) => void;

      /**
       * Boolean that indicates when the element used to trigger
       * `actions.router.set` is in the screen.
       */
      routeInView: boolean;

      /**
       * Boolean that indicates when the element used to trigger
       * `actions.source.fetch` is in the screen.
       */
      fetchInView: boolean;
    };

/**
 * Properties added by the {@link useInfiniteScroll} to the Router
 * state property.
 */
export type InfiniteScrollRouterState = {
  /**
   * The infinite scroll map of props.
   */
  infiniteScroll: {
    /**
     * Array of visited links.
     */
    links: string[];

    /**
     * Number of elements that will be automatically fetched.
     */
    limit: number;

    /**
     * List of pages by link. TODO: check this type.
     */
    pages: string[];

    /**
     * Archive (used by `usePostTypeInfiniteScroll`).
     */
    archive?: string;
  };

  /**
   * Any other prop.
   */
  [key: string]: unknown;
};

/**
 * The type of those packages the {@link useInfiniteScroll} hook depends on,
 * merged together.
 */
export type Packages = MergePackages<Source, Router>;
