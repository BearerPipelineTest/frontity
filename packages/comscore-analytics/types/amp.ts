import React from "react";
import { Package } from "frontity/types";

/**
 * Interface for AMP support.
 */
export default interface ComscoreAnalyticsAmp extends Package {
  /**
   * Root elements exported by this package.
   */
  roots: {
    /**
     * The comScore namespace.
     */
    comscoreAnalytics: React.ElementType;
  };

  /**
   * The state exposed by this package.
   */
  state: {
    /**
     * Comscore Analytics namespace.
     */
    comscoreAnalytics: {
      /**
       * Single tracking ID associated to a Comscore client.
       */
      trackingId?: string;

      /**
       * Array of tracking IDs associated to a Comscore client.
       */
      trackingIds?: string[];
    };
  };
}
