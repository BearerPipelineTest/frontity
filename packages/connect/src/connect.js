import { createContext, useContext, Component } from "react";
import { view } from "@frontity/react-easy-state";
import { warn } from "@frontity/error";

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext({});

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * A flag to know when a connected component is being rendered.
 */
let isConnected = false;

/**
 * The default options of connect.
 */
const defaultOptions = {
  injectProps: true,
};

const isStateless = (Comp) =>
  !(Comp.prototype && Comp.prototype.isReactComponent);

/**
 * Connects a React component with the Frontity connect store.
 *
 * @param Comp - The React component to be connected.
 * @param options - TO BE LINKED WHEN WE SWITCH TO TYPESCRIPT.
 *
 * @returns The same Component, but with the store included in the props.
 */
export const connect = (Comp, options) => {
  options = options ? { ...defaultOptions, ...options } : defaultOptions;

  if (isStateless(Comp)) {
    return view((props) => {
      const { state, actions, libraries } = useContext(context);
      isConnected = true;
      const rendered = Comp({
        ...props,
        ...(options.injectProps ? { state, actions, libraries } : {}),
      });
      isConnected = false;
      return rendered;
    });
  } else {
    if (options.injectProps) {
      /**
       * A wrapper for the Component that extracts the store from the context
       * and passes it to the Component.
       */
      class ConnectedComponent extends Component {
        /**
         * The render method of this React Class.
         *
         * @returns The same Component but with the store injected as props.
         */
        render() {
          const { state, actions, libraries } = this.context;
          const props = { ...this.props, state, actions, libraries };
          return <Comp {...props} />;
        }
      }
      ConnectedComponent.contextType = context;
      return view(ConnectedComponent);
    }
    return view(Comp);
  }
};

export const useConnect = () => {
  if (!isConnected)
    warn(
      "Warning: useConnect() is being used in a non connected component, " +
        "therefore the component won't update on state changes. " +
        "Please wrap your component with connect().\n"
    );

  return useContext(context);
};
