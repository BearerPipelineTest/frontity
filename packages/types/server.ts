import { Middleware, Next, ParameterizedContext } from "koa";
import { ResolveState, ResolveActions } from "./utils";
import Package from "./package";

/**
 * Frontity context for Koa middleware.
 */
export interface FrontityContext<Packages extends Package>
  extends ParameterizedContext {
  /**
   * Frontity namespace for custom properties.
   */
  frontity: any;

  /**
   * Packages state.
   */
  state: ResolveState<Packages["state"]>;

  /**
   * Packages actions.
   */
  actions: ResolveActions<Packages["actions"]>;

  /**
   * Packages libraries.
   */
  libraries: Packages["libraries"];

  /**
   * Packages server middleware.
   */
  server: Packages["server"];

  /**
   * Extended Koa context.
   */
  ctx: FrontityContext<Packages>;

  /**
   * Koa Next function.
   */
  next: Next;
}

/**
 * Types for server middleware in Frontity packages.
 */
export type Server<Packages extends Package> = Middleware<
  ResolveState<Packages["state"]>,
  FrontityContext<Packages>
>;

/**
 * Types for async server middleware in Frontity packages.
 */
export type AsyncServer<Packages extends Package> = Server<Packages>;
