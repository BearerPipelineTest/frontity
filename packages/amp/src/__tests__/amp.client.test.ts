import { createStore, InitializedStore } from "@frontity/connect";
import clone from "clone-deep";
import { Packages } from "../../types";
import Amp from "../client";

describe("AMP client hydration tests", () => {
  let store: InitializedStore<Packages>;

  beforeEach(() => {
    // Initialize the store
    store = createStore<Packages>(clone(Amp));

    // We have to mock these since the `createStore` it's not the one used
    // on the server and client.
    store.libraries = {
      source: {
        parse: jest.fn(),
        normalize: jest.fn(),
        stringify: jest.fn(),
      },
      frontity: {
        App: jest.fn(),
        render: jest.fn(),
        template: jest.fn(),
      },
    };
  });

  test("The initial App is overwritten with the Amp one", () => {
    const previousApp = store.libraries.frontity.App;
    store.actions.amp.beforeCSR();

    expect(store.libraries.frontity.App).not.toEqual(previousApp);
  });
});
