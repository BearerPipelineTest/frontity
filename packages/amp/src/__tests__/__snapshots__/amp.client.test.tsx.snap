// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AMP client hydration tests The \`App\` uses the CacheProvider 1`] = `
<Context.Provider
  value={
    Object {
      "insert": [Function],
      "inserted": Object {},
      "key": "css-amp",
      "nonce": undefined,
      "registered": Object {},
      "sheet": StyleSheet {
        "_insertTag": [Function],
        "before": null,
        "container": <head />,
        "ctr": 0,
        "isSpeedy": false,
        "key": "css-amp",
        "nonce": undefined,
        "prepend": undefined,
        "tags": Array [],
      },
    }
  }
>
  <mockConstructor />
</Context.Provider>
`;
